### DISCUSSION TITLE
Terminal Mode Keymaps

### DISCUSSION THREAD

**Discussion by Everduin94** [LazyVim/LazyVim#4092] [Q&A]
In a clean install of LazyVim, pressing `C-\` `C-n` exits terminal mode. However, if I try to remap `esc` or `caps-lock` to do this in terminal mode, it does not work.  The `set_terminal_keymaps` example below works in my original config, but not in LazyVim. Trying with `LazyVim.safe_keymap_set` does not work either.

```lua
local map = LazyVim.safe_keymap_set
function _G.set_terminal_keymaps()
  local opts = { noremap = true }
  vim.api.nvim_buf_set_keymap(0, "t", "<esc>", [[<C-\><C-n>]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-h>", [[<C-\><C-n><C-W>h]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-j>", [[<C-\><C-n><C-W>k]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-k>", [[<C-\><C-n><C-W>j]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-l>", [[<C-\><C-n><C-W>l]], opts)
end

vim.cmd("autocmd! TermOpen term://* lua set_terminal_keymaps()")
map("t", "<esc>", [[<C-\><C-n>]], { desc = "Escape Terminal" })
map("t", "<C-h>", [[<C-\><C-n><C-W>h]], { desc = "Move Panes" })
map("t", "<C-j>", [[<C-\><C-n><C-W>k]], { desc = "Move Panes" })
map("t", "<C-k>", [[<C-\><C-n><C-W>j]], { desc = "Move Panes" })
map("t", "<C-l>", [[<C-\><C-n><C-W>l]], { desc = "Move Panes" })
```

Terminal: WezTerm
OS: Arch Linux

**Comment by dpetka2001** [comment]
You only need to put this
```lua
function _G.set_terminal_keymaps()
  local opts = { noremap = true }
  vim.api.nvim_buf_set_keymap(0, "t", "<esc>", [[<C-\><C-n>]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-h>", [[<C-\><C-n><C-W>h]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-j>", [[<C-\><C-n><C-W>k]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-k>", [[<C-\><C-n><C-W>j]], opts)
  vim.api.nvim_buf_set_keymap(0, "t", "<C-l>", [[<C-\><C-n><C-W>l]], opts)
end

vim.cmd("autocmd! TermOpen term://* lua set_terminal_keymaps()")
```
in `~/.config/nvim/lua/config/autocmds.lua` I believe.

**Reply by Everduin94** [reply]
Doesn't work for me :(

Even tried `map("t", "<esc>", [[<C-\><C-n>]], { desc = "Escape Terminal" })` inside of the auto-command.

Also just for due diligence, I know my autocmds.lua is okay because I can print a message on terminal open.

It works with the literal `esc` key. But, not caps-lock. Which is remapped to `esc`, and works in all other context.

**Reply by dpetka2001** [reply]
You can't remap caps-lock to `<esc>` in the Neovim level. You should do that in your OS level.

Edit: actually how did you do that? Can you show the corresponding code of the remap?

**Reply by Everduin94** [reply]
Yup. - I didn't write enough detail. I have caps lock remapped to escape on the OS level via `setxkbmap` and `xcape`.

To be specific, I actually have short-press caps-lock mapped to escape, and long press caps-lock mapped to ctrl.

```sh
setxkbmap -option ctrl:nocaps
xcape -e 'Control_L=Escape'
```

In my [previous config](https://github.com/Everduin94/dotfiles/blob/main/nvim.bak/lua/user/terminal.lua) I used for 2-3 years before switching to LazyVim, the `set_terminal_keymaps` snippet I posted (+ my settings via `setxkbmap` and `xcape`) allowed me to press escape to exit terminal mode. - Hence why I'm posting here 😅.

**Reply by dpetka2001** [reply]
I have it configured like [this](https://github.com/dpetka2001/dotfiles/blob/19c578e8e57a1079c3ca9d7a9efff38712b78967/dot_config/nvim/lua/config/autocmds.lua#L44-L62) and it works. I haven't tried the way you described to remap `Caps` to `<esc>`. I use Linux Mint 21.3 and made the remap via `Keyboard -> Layouts -> Options -> Caps Lock behavior -> Swap Esc and Caps Lock` and it works as intended for me. It has something to do specific with your system configuration I believe.
